### PR TITLE
Simplify sorted top docs merging in SearchPhaseController

### PR DESCRIPTION
**Pull request by s1monw**
Today we have several code paths to merge top docs based on the number of
search results returned from the shards. If there is a only a single shard
holding any hits we go a different code path with quite some complexity while
if there are more than one the code is basically duplicated to safe the
creation of a dense array of top docs which can be large if there are many results.
This commit removes the need of the dense array and in-turn the justification for
the optimization. This commit introduces a single code path to merge top docs.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 1]
```diff
@@ -147,156 +146,105 @@ private static long optionalSum(long left, long right) {
...
-                }
-                docs = new ScoreDoc[resultDocsSize + numSuggestDocs];
-                for (int i = 0; i < resultDocsSize; i++) {
-                    ScoreDoc scoreDoc = scoreDocs[offset + i];
-                    scoreDoc.shardIndex = shardIndex;
-                    docs[i] = scoreDoc;
-                    docsOffset++;
+        if (size != -1) {
+            final ScoreDoc[] mergedScoreDocs = topDocs.isEmpty() ? EMPTY_DOCS :
+                mergeTopDocs(topDocs, size, ignoreFrom ? 0 : from).scoreDocs;
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: Maybe we should still apply the single shard optim here by not calling mergeTopDocs if `topDocs.size() ==  1` ?

**Review by jimczi — APPROVED** [review]
The change looks great ! 
I left a minor comment regarding the single shard optim.

**Review by jpountz — APPROVED** [review]
LGTM

**Comment by s1monw** [comment]
@jpountz @jimczi I pushed another change and I am not sure we should keep the optimization to be honest...

**Comment by jimczi** [comment]
>  I am not sure we should keep the optimization to be honest...

It's not a killer optim for sure and cases that would benefit from it are edge cases that we don't want to handle (returning a lot of documents from a single shard) so you can remove it if you want.
If we really want it we could also make the optim directly in Lucene TopDocs.mergeTopDocs which does not check the size of the input array.

**Comment by s1monw** [comment]
@jimczi I simplified the scope of the optimization to 1 shard and no pagination. I think this is worth it and a good compromise...

**Pull request merged**